notes.lua

```diff
@@ -0,0 +1,324 @@
+--@ module = true
+
+local gui = require('gui')
+local widgets = require('gui.widgets')
+local textures = require('gui.textures')
+local overlay = require('plugins.overlay')
+local guidm = require('gui.dwarfmode')
+local text_editor = reqscript('internal/journal/text_editor')
+
+local green_pin = dfhack.textures.loadTileset('hack/data/art/note-green-pin.png', 32, 32, true)
+
+NotesOverlay = defclass(NotesOverlay, overlay.OverlayWidget)
+NotesOverlay.ATTRS{
+    desc='Render map notes.',
+    viewscreens='dwarfmode',
+    default_enabled=true,
+    overlay_onupdate_max_freq_seconds=30,
+}
+
+local waypoints = df.global.plotinfo.waypoints
+local map_points = df.global.plotinfo.waypoints.points
+
+function NotesOverlay:init()
+    self.notes = {}
+    self.note_manager = nil
+    self.last_click_pos = {}
+    self:reloadVisibleNotes()
+end
+
+function NotesOverlay:overlay_onupdate()
+    self:reloadVisibleNotes()
+end
+
+function NotesOverlay:overlay_trigger(args)
+    return self:showNoteManager()
+end
+
+function NotesOverlay:onInput(keys)
+    if keys._MOUSE_L then
+        local top_most_screen = dfhack.gui.getDFViewscreen(true)
+        if dfhack.gui.matchFocusString('dwarfmode/Default', top_most_screen) then
+            local pos = dfhack.gui.getMousePos()
+
+            local note = self:clickedNote(pos)
+            if note ~= nil then
+                self:showNoteManager(note)
+            end
+        end
+    end
+end
+
+function NotesOverlay:clickedNote(click_pos)
+    local pos_curr_note = same_xyz(self.last_click_pos, click_pos)
+        and self.note_manager
+        and self.note_manager.note
+        or nil
+
+    self.last_click_pos = click_pos
+
+    local last_note_on_pos = nil
+    local first_note_on_pos = nil
+    for _, note in ipairs(self.notes) do
```

Comment:
this is now O(N) in the number of notes -- *probably* ok for likely use cases. I just get nervous when I see O(N) algorithms for (technically) unbounded N

Reply:
but `self.notes` is only notes currently visible on the screen, so would be really hard to make it slow IMO.
I will improve the var name to make it obvious.

Reply:
we could consider optimizing `:reloadVisibleNotes` function if you have an idea how to do this.
there, I believe, is weak point which can make e.g. map navigation slow if there will really big # of notes.

Reply:
we can stress test this and see if we need to make any further optimizations